0.6.0 release candidate

dmolesUC opened the pull request:
- Move Alma configuration out of `AlmaBase` into scripts
- Add README section on Alma configuration
- Misc. code cleanup:
  - remove unnecessary explicit module references when we're already inside the namespace (`BerkeleyLibrary::TIND::Mapping::AlmaBase.collection_parameter_hash` ⇒ `AlmaBase.collection_parameter_hash`, etc.)
  - rename `ExternalTindField.tind_fields_from_alma_id` to `tind_mms_id_fields`, and remove the second `id` parameter which was only used to log a warning if `mms_id` was nil (warning is now logged from `AlmaBase.tind_record` instead)
- Update changelog and gem version

yzhoubk commented:
Others look good, please see comments on this one:
. rename ExternalTindField.tind_fields_from_alma_id to tind_mms_id_fields, and remove the second id parameter which was only used to log a warning if mms_id was nil (warning is now logged from AlmaBase.tind_record instead)

Could we log a nil mms_id in 'BerkeleyLibrary::Alma::RecordId' ?   What 'AlmaBase.tind_record' get will be a nil Alma_record.

dmolesUC commented:
@yzhoubk `BerkeleyLibrary::Alma::RecordId` doesn't log a warning. My thinking is that it's a low-level class and if an application wants to log a warning or raise an error on a nil ID, the application should check the return value.

I added a test to make sure the mapping still works if there's no 001 field, but I'm wondering if that's actually a good idea. For instance, we probably don't want to send a record to TIND if we can't create an 035 field for it. I think some of these cases where we're logging a warning, it might be better for the mapping code to raise an error, and then the top-level code can catch it, log the error, and skip that record. At least if it's something that can't easily be corrected later. Likewise, if the collection parameters are bad, that probably means the whole batch is bad. What do you think?

---

(As a side note -- I noticed we were still writing an 035 even if `is_035_from_mms_id` was true and there was no MMS ID -- we would just write, for instance,

```xml
<datafield tag="035" ind1=" " ind2=" ">
  <subfield code="a">(pre1912)</subfield>
</datafield>
```

I changed that so we would only add the 035 if the MMS ID was not nil.)

yzhoubk commented:
The mms_id for deriving 035 field is taken from an Alma record.  In real case,  if mms_id is nil, there will be no Alma record, if there is no Alma record, the mapping will do nothing,  code will never reach the point to create a 035 field for nil mms_id. 

That's good idea to raise errors.  I am going to raise error for a nil  alma record,  nil (mms_id, oskicat No, Barcode), invalid collection information.  I'll check other cases. Thanks!

yzhoubk commented:
I notice you have a release branch. If it is convenient for you, could you please add  raise error to  nil  alma record,  nil id and invalid collection information?   Or please let me know how to update the branch?

dmolesUC commented:
@yzhoubk Since the MMS ID that we use to generate the 035 is taken from the 001, it's possible (if not at all likely) that it's not the same as the ID we use to look up the record. It's also possible (but even more unlikely, since something would have to be really broken on the Alma side) that we might get a record with no 001 field. So I think the chance of having a nil MMS ID at the time we generate the 035 is very low, but it's better to fail quickly in that case.

If you want to create another branch of your own, based on the 0.6.0-rc branch here, and add the checks to that branch, I can merge your changes into this one. Or I can take care of it -- just let me know if you think of any other cases. Sorry I got pulled into something else and probably won't get to it today, but I should have time tomorrow.

yzhoubk commented:
I've created a branch locally base on the upstream 0.6.0, I'll look into workflow, add error raises, including 035 field, update test

yzhoubk commented:
@dmolesUC 

I've added error raise for 

1. nil mms_id, 
2. nil Alma_record,
3. unqualified Alma, 
4. unqualified collection information, 
5. nil BarCode or RecordId gotten from Alma module

Originally I thought methods in 'Alma.rb',  'field_catalog_util.rb' files have been tested in other modules, so did not create RSpec files for them.

Now I added their RSpec files 'Alma_spec.rb', 'field_catalog_util_spec.rb';  

But I run into an issue with field_catalog_util_spec.rb': it complains about "no method" when a method from 'Misc'  is called by a method in module 'FieldCatalogUtil'.  I use a dummy objet which extending module, I did this way to test other modules and did not have any problem. (testing is commented out, not using x)

Could you help take a look? Or should I just skip on testing  'FieldCatalogUtil',  it definitely has been tested in other module. 

This is the branch:
https://github.com/yzhoubk/tind/tree/Add_error_raise_0.6.0-rc

yzhoubk commented:
@dmolesUC 

I include other modules through a dummy class, it seems work now. This is not nice looking, if there are dozens of modules, all of them will have to be listed.

class DummyFieldCatalogUtil
            include Misc
            include FieldCatalogUtil
            include TindSubfieldUtil
          end

dmolesUC commented:
@yzhoubk That shouldn't be necessary. If `FieldCatalogUtil` can't work without methods from `Misc` and `TindSubfieldUtil`, then `FieldCatalogUtil` should probably include `Misc` explicitly. (That will also simplify the classes that include `FieldCatalogUtil` as they no longer have to include the other modules.)
